File: api_service/api/forum.py
from http import HTTPStatus
from typing import List

from fastapi import APIRouter, Depends, Response

from api.dependency import filter_parameters, filter_users_parameters
from api.exceptions import HttpNotFoundException
from models.forum import Forum
from models.thread import Thread
from models.user import User
from services.forum import forum_service

router = APIRouter()


@router.post('/create', status_code=201)
async def create_forum(
        item: Forum,
        response: Response
):
    forum, not_unique = await forum_service.create_forum(item)

    if not forum:
        raise HttpNotFoundException()
    if not_unique:
        response.status_code = HTTPStatus.CONFLICT

    return forum


@router.get('/{slug}/details')
async def get_forum_details(
        slug: str
) -> Forum:
    forum = await forum_service.get_by_slug(slug)

    if not forum:
        raise HttpNotFoundException()

    return forum


@router.post('/{slug}/create', status_code=201)
async def create_thread(
        slug: str,
        item: Thread,
        response: Response
):
    if not item.forum:
        item.forum = slug
    thread, not_unique = await forum_service.create_thread(item)

    if not thread:
        raise HttpNotFoundException()
    if not_unique:
        response.status_code = HTTPStatus.CONFLICT

    return thread


@router.get('/{slug}/users')
async def get_forum_users(
        slug: str,
        filter_params: dict = Depends(filter_users_parameters)
) -> List[User]:
    users = await forum_service.get_forum_users(
        slug=slug,
        filter_params=filter_params
    )

    if not isinstance(users, list):
        raise HttpNotFoundException()

    return users


@router.get('/{slug}/threads')
async def get_forum_threads(
        slug: str,
        filter_params: dict = Depends(filter_parameters)
) -> List[Thread]:
    threads = await forum_service.get_forum_threads(
        slug=slug,
        filter_params=filter_params
    )

    if not isinstance(threads, list):
        raise HttpNotFoundException()

    return threads

File: api_service/api/forum.go
package api

import (
	"db_tp/models"
	"db_tp/services"
	"fmt"
	"github.com/mailru/easyjson"
	"github.com/valyala/fasthttp"
	"strconv"
	"time"
)

func CreateForum(ctx *fasthttp.RequestCtx) {
	item := &models.Forum{}
	easyjson.Unmarshal(ctx.PostBody(), item)

	forum, notUnique := services.ForumSrv.CreateForum(item)
	if forum == nil {
		errMsg := models.ErrMsg{Message: fmt.Sprintf("Can't find user with nickname:  %s", item.User)}
		response, _ := easyjson.Marshal(errMsg)
		ctx.SetBody(response)
		ctx.SetStatusCode(404)
		ctx.SetContentType("application/json")
		return
	}
	if notUnique {
		errMsg := models.ErrMsg{Message: fmt.Sprintf("Not unique fields")}
		response, _ := easyjson.Marshal(errMsg)
		ctx.SetBody(response)
		ctx.SetStatusCode(409)
		ctx.SetContentType("application/json")
		return
	}

	resp, _ := easyjson.Marshal(forum)
	ctx.Response.SetBody(resp)
	ctx.SetContentType("application/json")
	ctx.Response.SetStatusCode(201)
}

func GetForumDetails(ctx *fasthttp.RequestCtx) {
	slug := ctx.UserValue("slug").(string)

	forum := services.ForumSrv.GetBySlug(slug)

	if forum == nil {
		errMsg := models.ErrMsg{Message: fmt.Sprintf("Can't find forum with slug:  %s", slug)}
		response, _ := easyjson.Marshal(errMsg)
		ctx.SetBody(response)
		ctx.SetStatusCode(404)
		ctx.SetContentType("application/json")
		return
	}

	resp, _ := easyjson.Marshal(forum)
	ctx.Response.SetBody(resp)
	ctx.SetContentType("application/json")
	ctx.Response.SetStatusCode(200)
}

func CreateThread(ctx *fasthttp.RequestCtx) {
	slug := ctx.UserValue("slug").(string)
	item := &models.Thread{}
	easyjson.Unmarshal(ctx.PostBody(), item)
	if item.Forum == "" {
		item.Forum = slug
	}

	thread, notUnique := services.ForumSrv.CreateThread(item)
	if thread == nil {
		errMsg := models.ErrMsg{Message: fmt.Sprintf("Can't find")}
		response, _ := easyjson.Marshal(errMsg)
		ctx.SetBody(response)
		ctx.SetStatusCode(404)
		ctx.SetContentType("application/json")
		return
	}
	if notUnique {
		errMsg := models.ErrMsg{Message: fmt.Sprintf("Not unique fields")}
		response, _ := easyjson.Marshal(errMsg)
		ctx.SetBody(response)
		ctx.SetStatusCode(409)
		ctx.SetContentType("application/json")
		return
	}

	resp, _ := easyjson.Marshal(thread)
	ctx.Response.SetBody(resp)
	ctx.SetContentType("application/json")
	ctx.Response.SetStatusCode(201)
}

func GetForumUsers(ctx *fasthttp.RequestCtx) {
	slug := ctx.UserValue("slug").(string)
	desc, _ := strconv.ParseBool(string(ctx.QueryArgs().Peek("desc")))
	limitStr := string(ctx.QueryArgs().Peek("limit"))
	var limit int
	if limitStr == "" {
		limit = 100
	} else {
		limit, _ = strconv.Atoi(limitStr)
	}
	since := string(ctx.QueryArgs().Peek("since"))

	users := services.ForumSrv.GetForumUsers(slug, desc, limit, since)

	if users == nil {
		errMsg := models.ErrMsg{Message: fmt.Sprintf("Can't find users for forum with slug:  %s", slug)}
		response, _ := easyjson.Marshal(errMsg)
		ctx.SetBody(response)
		ctx.SetStatusCode(404)
		ctx.SetContentType("application/json")
		return
	}

	resp, _ := easyjson.Marshal(users)
	ctx.Response.SetBody(resp)
	ctx.SetContentType("application/json")
	ctx.Response.SetStatusCode(200)
}

func GetForumThreads(ctx *fasthttp.RequestCtx) {
	slug := ctx.UserValue("slug").(string)
	desc, _ := strconv.ParseBool(string(ctx.QueryArgs().Peek("desc")))
	limitStr := string(ctx.QueryArgs().Peek("limit"))
	var limit int
	if limitStr == "" {
		limit = 100
	} else {
		limit, _ = strconv.Atoi(limitStr)
	}

	var threads models.Threads

	since := string(ctx.QueryArgs().Peek("since"))
	if since != "" {
		sinceTime, _ := time.Parse(time.RFC3339, since)
		threads = services.ForumSrv.GetForumThreads(slug, desc, limit, &sinceTime)
	} else {
		threads = services.ForumSrv.GetForumThreads(slug, desc, limit, nil)
	}

	if threads == nil {
		errMsg := models.ErrMsg{Message: fmt.Sprintf("Can't find threads for forum with slug:  %s", slug)}
		response, _ := easyjson.Marshal(errMsg)
		ctx.SetBody(response)
		ctx.SetStatusCode(404)
		ctx.SetContentType("application/json")
		return
	}

	resp, _ := easyjson.Marshal(threads)
	ctx.Response.SetBody(resp)
	ctx.SetContentType("application/json")
	ctx.Response.SetStatusCode(200)
}
